### DISCUSSION TITLE
Question: Build dependencies vs. runtime dependencies?

### DISCUSSION THREAD

**Discussion by benblasco** [librespot-org/librespot#768] [General]
Hi team,

Firstly, thank you to all the contributors who have made a great tool for people to consume.  I am really grateful to be able to use librespot.  

My question today is:
What are the _build_ dependencies for librespot as compared to the _run_ dependencies?

I see the following listed for Fedora under https://crates.io/crates/librespot
- alsa-lib-devel
- make
- gcc

Then of course one needs to install
- cargo (which will install rust as a dependency

It is not clear to me from the docs which packages are needed just for the build, and which are needed just for execution of the compiled binary.  Is there any chance you can point me in the right direction?

The reason I want to do this is to be able to run a build container (with podman, not docker), extract the compiled artifact, and then insert it into a lighter runtime container to actually run the application.  I will share my work on this separately.

Thanks to anybody who can reply with their insights!

**Comment by JasonLG1979** [comment]
`librespot` compiled as a standalone binary AFAIK doesn't have any runtime dependencies other than a working audio system. So if you compiled it with the ALSA backend (like it looks like you're wanting to do?) all you'd need to run it is a system with working ALSA audio (or something compatible with ALSA like PulseAudio or PipeWire). Your mileage may vary but `librespot` compiled with the ALSA backend *Just Works* for me on my Fedora 34 desktop system with PipeWire without any issues.

**Reply by JasonLG1979** [reply]
The only caveat on a system running PulseAudio and PipeWire being that you can't directly address the audio hardware because PulseAudio and PipeWire already have them open.

**Reply by benblasco** [reply]
Hi Jason, sorry I never saw your reply! I'm a fedora user too and have been running it in a self built fedora 32 container (running on fedora 32 server) which I am about to update to fedora 35, and switching over to pipewire. I'll be sharing my findings with the project when I do!

**Reply by JasonLG1979** [reply]
> Hi Jason, sorry I never saw your reply! I'm a fedora user too and have been running it in a self built fedora 32 container (running on fedora 32 server) which I am about to update to fedora 35, and switching over to pipewire. I'll be sharing my findings with the project when I do!

By default on basically all distros ALSA, Pulse Audio and Pipewire have "default" virtual devices that are software mixed. On a system with Pipewire you can address the hardware directly if you like but only if nothing else is using it at the moment and nothing else will be able to use it while librespot is.

**Reply by benblasco** [reply]
I think it has been crashing the application on occasion because of what you have said about addressing hardware directly, as I am just exposing /dev/snd to my container.  I will change this when I move to Fedora 35 and try out Pipewire.  Once I have this working I will share my work for those who are interested.
